Feature/account crud

HyunJng opened the pull request:
**[기존 의견 반영하지 못한 것]**
지출, 수입에 대한 카테고리를 분리해서 enum클래스를 만들지 못했습니다. 
- 여러 Enum타입을 Interface로 묶어 Entity의 필드로 넣는 것은 불가능한 일이였습니다. (혹시 관련된 글 읽고싶으시면 https://stackoverflow.com/questions/54093620/jpa-enum-inheritance)
- 그래서 AccountCategory라는 하나의 enum클래스에서 관리하되, 둘을 구별해주는 필드를 추가해두었습니다.
- 만약 각 타입(수입/지출)의 카테고리 리스트를 분리해서 보고 싶다면 바로 꺼내볼 수 있게 Category enum클래스의 field로 참조할 수 있도록 해두었습니다.

HyunJng committed:
feat([Daywork]): Daywork create기능 추가

[변경 파일 설명] * 생성 파일은 설명 제외.
- ScheduleController : 기존에 다른 분이 만들어둔
getSchedule()코드에서 오류나길래 일시적으로 return null처리

- domain
: 전체적으로 생성자 접근제어 추가
	. DateTime
	. Daywork : status에 default = NOT_COMPLETE 추가
	. Schedule: budget에 default = 0 추가

HyunJng committed:
feat([Daywork]): Daywork update기능 추가

update기능 추가. 추가 외의 변경(삭제) 사항 없음.

HyunJng committed:
fix([Schedule]): Schedule 조회시 User정보 참조하도록 수정

- Service에서 LoginUser정보 생성하고 Repository로 넘김
- Repository의 Query문 수정

HyunJng committed:
feat([Daywork]): Daywork 삭제 기능 추가

HyunJng committed:
refactor([Daywork]): Daywork 관련 feedback 반영

1. DayworkService -> ScheduleService 참조관계에서 ScheduleService ->
DayworkService 참조관계로 변경
2. Repository에서 외래키에 대한 parameter를 객체로 받는 것이 아니라
ID로 받도록 수정
 - 단, exists___AND___로 하는 것은 반영하지 않음. (ID로 검색할 시 오류남)
3. 클래스와
메서드 이름 명확하게 용도 알 수 있도록 변경

HyunJng committed:
feat([Account]): Account  Create 기능 추가

[저번 회의 내용 중 반영하지 못한 것]
- Account의 Category Enum타입을 Interface로 묶어서 사용하는
것은 불가능
	=> 대신 하나의 Enum을 사용하되, type에 대한 field 정보를 추가해 구분할 수 있도록 구현

HyunJng committed:
feat([Account]): Account  Update 기능 추가

HyunJng committed:
feat([Account]): Account  Delete 기능 추가

- 해당 Schedule위에 Account, Daywork 모두 존재하지 않아야 Schedule이 삭제되도록
deleteDaywork도 같이 수정

JinhyeonKwak commented on src/main/java/com/whatachad/app/api/ScheduleCrudApi.java:
```diff
@@ -0,0 +1,92 @@
+            @ApiResponse(responseCode = "405", description = "Method Not Allowed")
+    })
+    @PutMapping("/{YYYYMM}/dayworks/{DD}/{daywork_id}")
+    public ResponseEntity<UpdateDayworkResponseDto> editDaywork(@RequestBody UpdateDayworkRequestDto requestDto, @PathVariable Long daywork_id);
```
제가 PathVariable을 url에서 정의할 때 "daywork_id"같은 형식으로 정의했지만, 이건 프론트 분들과 소통하기 위한 임의 변수일 뿐이라서 백엔드 내부에서는 java convention을 지키는 게 좋을 거 같아요. 원래 하셨던 것처럼 camel case 형태로 나타내는 것 말이죠.

HyunJng replied:
넵 조언 감사합니당!!!

JinhyeonKwak commented on src/main/java/com/whatachad/app/api/ScheduleCrudApi.java:
```diff
@@ -0,0 +1,92 @@
+            @ApiResponse(responseCode = "405", description = "Method Not Allowed")
+    })
+    @PostMapping("/{YYYYMM}/dayworks/{DD}")
+    public ResponseEntity<CreateDayworkResponseDto> registerDaywork(@RequestBody CreateDayworkRequestDto requestDto, @PathVariable("YYYYMM") String ym, @PathVariable("DD") Integer date);
```
"ym"
저는 알고리즘 문제 풀 때를 제외하고는 변수 정의할 때 줄임말을 안 쓰려고 노력하고 있어요.
그래서 최대한 변수명이나 메서드명만으로도 코드를 이해할 수 있도록 하지만 항상 어려운 일인 것 같네요.
변수명이나 메서드명이 길어져서 보기 힘들어지면 개행으로 정리를 하거나, 최대한 의미가 전달되도록 줄여서 씁니다.

HyunJng replied:
이것도 유의하겠습니다!

JinhyeonKwak commented on src/main/java/com/whatachad/app/controller/ScheduleCrudController.java:
```diff
@@ -0,0 +1,102 @@
+        boolean remain = scheduleService.existsAnythingOnSchedule(schedule.getId());
+        if(!remain){
+            scheduleService.deleteSchedule(schedule.getId());
+        }
```
daywork를 지울 때 schedule을 지우는 이유가 뭐 때문일까요? 이 로직에 의하면 schedule에 daywork 혹은 account가 하나도 남아 있지 않을 때 같이 지우는 것 같은데,  그렇다면 나중에 daywork나 account를 생성할 때 schedule을 다시 생성해야 하는 문제가 발생할 것 같다는 생각이 들어요.
제가 잘못 이해한 게 있다면 설명 부탁드려도 될까요?

HyunJng replied:
이 때 같이 지우지 않는다면 schedule을 지울 타이밍 없이 계속 쌓이기만 할 것 같아서요. 연관된 Daywork와 account이 없으면 schedule은 조회될 일이 없는 데이터일텐데 db에 남겨두는건 공간만 차지할 것이라 생각해서 삭제했습니다. 혹시 남겨두는 편이 좋을까요?

JinhyeonKwak commented on src/main/java/com/whatachad/app/service/ScheduleService.java:
```diff
@@ -0,0 +1,108 @@
+    public boolean existsAnythingOnSchedule(Long schedule_id) {
+        boolean existsDaywork = dayworkService.isDayworkBySchedule(schedule_id);
+        boolean existsAccount = accountService.isAccountBySchedule(schedule_id);
+
```
이 메서드의 용도가 존재하는 스케줄에 daywork나 account가 있는지 확인하는 것이겠죠?
그렇다면 controller에서 코멘트한 바와 같이 이 메서드의 용도가 잘 이해가 안 되네요.
제가 놓치고 있는 부분이 있는 걸까요?

HyunJng replied:
위에 작성했습니다!!

JinhyeonKwak commented on src/main/java/com/whatachad/app/type/AccountCategory.java:
```diff
@@ -0,0 +1,42 @@
+    public static AccountCategory valueOfLabel(String label) {
+        return BY_LABEL.get(label);
+    }
+
```
오! 멋지네요!! 한 수 배우고 갑니다 ㅎㅎ

HyunJng replied:
뿌듯ㅎㅎ 감사합니다😊

HyunJng committed:
refactor([Account]): Account  피드백 반영

- 일부 변수명 Camel case로 변경, 직관적인 변수명으로 수정
- daywork와 account 삭제 시,
schedule은 남겨두도록 로직 변경
